docs/concepts/plugins.md

```diff
@@ -0,0 +1,235 @@
+but still make sure they are only executed for specific commands.
+
+You can also nest commands, e.g. if you want to provide a command `serverless deploy single` you can simply nest
+commands as defined in the following example. Those nested commands have their own lifecycle events and do not inherit
+them from their parents.
+
+```javascript
+constructor() {
+  this.commands = {
+  deploy: {
+    lifecycleEvents: [
+      'resources',
+      'functions'
+    ],
+    commands: {
+      single: {
+        lifecycleEvents: [
+          'resources',
+          'functions'
+        ],
+      },
+    },
+  },
+};
+```
+
+### Hook
+
+Hooks allow to connect specific lifecycle events to functions in a Plugin. In the constructor of your class you define
+a *hooks* variable that the Plugin System will use once a specific command is running. Any hook can bind to any
+lifecycle event from any command, not just from commands that the same plugin provides.
+
+This allows to extend any command with additional functionality.
+
+```javascript
+'use strict';
+
+class Deploy {
+  constructor() {
+    this.commands = {
+      deploy: {
+        lifecycleEvents: [
+          'resources',
+          'functions'
+        ]
+      },
+    };
+
+    this.hooks = {
+      'before:deploy:resources': this.beforeDeployResources,
+      'deploy:resources': this.deployResources,
+      'after:deploy:functions': this.afterDeployFunctions
+    };
+  }
+
+  beforeDeployResources() {
+    console.log('Before Deploy Resources');
+  }
+
+  deployResources() {
+    console.log('Deploy Resources');
+  }
+
+  afterDeployFunctions() {
+    console.log('After Deploy Functions');
+  }
+}
+
+module.exports = Deploy;
+```
+
+### Options
+
+Each (sub)command can have multiple options. Options are passed with a single dash (`-`) or a double dash (`--`) like
+this: `serverless function deploy --function functionName` or `serverless resource deploy -r resourceName`.
+The `options` object will be passed in as the first parameter to your hook function where you can access it.
+
+```javascript
+'use strict';
+
+class Deploy {
+  constructor() {
+    this.commands = {
+      deploy: {
+        lifecycleEvents: [
+          'functions'
+        ],
+        options: {
+          function: {
+            usage: 'Specify the function you want to deploy (e.g. "--function myFunction")'
+          }
+        }
+      },
+    };
+
+    this.hooks = {
+      'deploy:functions': this.deployFunction
+    }
+  }
+
+  deployFunction(options) {
+    console.log('Deploying function: ', options.function);
+  }
+}
+```
+
+Options can be accessed with the help of the `options` object which will be automatically passed as the first parameter
+of the corresponding hook function.
+
+## Registering plugins for hooks and commands
+
+A user has to define the plugins they want to use in the serverless.yaml. We do not auto-detect plugins from installed
```

Comment:
Can we add an example of the syntax in serverless.yml to make this clearer for users